libs/@local/hash-validation/src/data_type.rs

```diff
@@ -58,31 +59,65 @@ pub enum DataTypeConstraint {
         actual: JsonValue,
         expected: JsonValue,
     },
-    #[error("the provided value is not greater than or equal to the minimum value")]
+    #[error("the provided value is not one of the expected values")]
+    Enum {
+        actual: JsonValue,
+        expected: JsonValue,
+    },
+    #[error(
+        "the provided value is not greater than or equal to the minimum value, got `{actual}`, \
+         expected `{expected}`"
+    )]
     Minimum {
         actual: JsonValue,
         expected: JsonValue,
     },
-    #[error("the provided value is not less than or equal to the maximum value")]
+    #[error(
+        "the provided value is not less than or equal to the maximum value, got `{actual}`, \
+         expected `{expected}`"
+    )]
     Maximum {
         actual: JsonValue,
         expected: JsonValue,
     },
-    #[error("the provided value is not greater than the minimum value")]
+    #[error(
+        "the provided value is not greater than the minimum value, got `{actual}`, expected \
+         `{expected}`"
+    )]
     ExclusiveMinimum {
         actual: JsonValue,
         expected: JsonValue,
     },
-    #[error("the provided value is not less than the maximum value")]
+    #[error(
+        "the provided value is not less than the maximum value, got `{actual}`, expected \
+         `{expected}`"
+    )]
     ExclusiveMaximum {
         actual: JsonValue,
         expected: JsonValue,
     },
-    #[error("the provided value is not a multiple of the expected value")]
+    #[error(
+        "the provided value is not a multiple of the expected value, got `{actual}`, expected \
+         `{expected}`"
+    )]
     MultipleOf {
         actual: JsonValue,
         expected: JsonValue,
     },
+    #[error(
+        "the provided value is not longer than the minimum length, got `{actual}`, expected \
+         `{expected}`"
+    )]
+    MinLength { actual: String, expected: usize },
+    #[error(
+        "the provided value is not shorter than the maximum length, got `{actual}`, expected \
+         `{expected}`"
```

Comment:
'not longer' implies an exclusive minimum
```suggestion
        "the provided value is shorter than the minimum length, got `{actual}`, expected \
         a string of at least length `{expected}`"
    )]
    MinLength { actual: String, expected: usize },
    #[error(
        "the provided value is longer than the maximum length, got `{actual}`, expected \
         a string of at most length `{expected}`"
```